### PR TITLE
Implement GUI for API keys

### PR DESCRIPTION
**Pull request by blokhin**

### CODE REVIEW

File: routes/[version]/apikey.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+const { checkAuth } = require('../../middlewares/auth');
+const { selectAPIToken, setAPIToken, removeAPIToken } = require('../../services/db');
+
+module.exports = {
+    get: [checkAuth, get],
+    put: [checkAuth, put],
+    delete: [checkAuth, del],
+};
+
+function generateAPIToken() {
```
Comment (knopki) [inline routes/[version]/apikey.js:10]: `Math.random()` is not ok for cryptography. `crypto` is ok. I propose this alternative:

```js
const generateAPIToken = (length = 64) => 'METIS' + crypto.randomBytes(length).toString("base64");
```

File: routes/[version]/apikey.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * @api {del} /v0/apikey Remove API key
+ * @apiName RemoveAPIKey
+ * @apiGroup Users
+ * @apiDescription Remove all API keys
+ * @apiPermission GUI_ONLY
+ */
+async function del(req, res, next) {
+    removeAPIToken(req.user.id);
+    return res.json({});
```
Comment (knopki) [inline routes/[version]/apikey.js:43]: Maybe `{ apikey: null }` with response code 200 or just nothing with response code 204? I don't like the fact that the data structure is changing.

File: routes/[version]/apikey.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * @api {get} /v0/apikey Check if API key exists
+ * @apiName GetAPIKey
+ * @apiGroup Users
+ * @apiDescription Get API key, if any
+ * @apiPermission GUI_ONLY
+ */
+async function get(req, res, next) {
+    return res.json({ apikey: await selectAPIToken(req.user.id) });
```
Comment (knopki) [inline routes/[version]/apikey.js:54]: I think this is insecure. I like to return nothing with response code 204 if set and nothing with response code 404 if not set.

**Review by knopki — CHANGES_REQUESTED** [review]
Let's fight in the chat room.

**Comment by knopki** [comment]
Why is SSE not used? :trollface:

File: routes/[version]/apikey.js (hunk 1 above)
Comment (pyoner) [reply routes/[version]/apikey.js:10]: @blokhin don't reinvent the wheel
**uuid** is enough or use another library :)

File: routes/[version]/apikey.js (hunk 2 above)
Comment (pyoner) [reply routes/[version]/apikey.js:43]: I vote for 204 code, more info [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE)

A [202](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202) (Accepted) status code if the action will likely succeed but has not yet been enacted.
A [204](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204) (No Content) status code if the action has been enacted and no further information is to be supplied.
A [200](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200) (OK) status code if the action has been enacted and the response message includes a representation describing the status.

File: routes/[version]/apikey.js (hunk 1 above)
Comment (knopki) [reply routes/[version]/apikey.js:10]: Yes, `crypto.randomUUID()` is nice too. And available without external dependency.

**Comment by blokhin** [comment]
> Why is SSE not used? :trollface:

because this is not a server-originated event, it's fully controlled by a client

File: routes/[version]/apikey.js (hunk 3 above)
Comment (blokhin) [reply routes/[version]/apikey.js:54]: Why insecure? This endpoint requires authorization.

Comment (knopki) [reply routes/[version]/apikey.js:54]: API key is not scoped (=full control). API key doesn't have any expiration date. Also, if it is possible to show the token some time after creation, then everyone will immediately realize that the tokens are stored in the database unencrypted. Such a shame.
This product doesn't seem to require any particularly high security requirements, but just not giving the token away unnecessarily isn't difficult either.

File: routes/[version]/apikey.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+const { checkAuth } = require('../../middlewares/auth');
+const { selectAPIToken, setAPIToken, removeAPIToken } = require('../../services/db');
+
+module.exports = {
+    get: [checkAuth, get],
+    put: [checkAuth, put],
+    delete: [checkAuth, del],
+};
+
+const generateAPIToken = (length = 11) => 'METIS_' + crypto.randomBytes(length).toString("hex"); // 28 chars
```
Comment (blokhin) [inline routes/[version]/apikey.js:11]: @knopki @pyoner I use this in favor of UX: the key should not be too long plus it should contain alphanumeric symbols only in order to be copied with the mouse easily (double click copies the entire string, not the part of it)

File: routes/[version]/apikey.js (hunk 3 above)
Comment (blokhin) [reply routes/[version]/apikey.js:44]: Created #88 to keep up with this in future (we don't have time for this now)

**Pull request merged**